p2p/stream/protocols/sync/client.go

```diff
@@ -0,0 +1,344 @@
+	bnsStr := strings.Join(ss, ",")
+	return fmt.Sprintf("REQUEST [GetBlockHashes: %s]", bnsStr)
+}
+
+func (req *getBlockHashesRequest) IsSupportedByProto(target sttypes.ProtoSpec) bool {
+	return target.Version.GreaterThanOrEqual(MinVersion)
+}
+
+func (req *getBlockHashesRequest) Encode() ([]byte, error) {
+	msg := syncpb.MakeMessageFromRequest(req.pbReq)
+	return protobuf.Marshal(msg)
+}
+
+func (req *getBlockHashesRequest) getHashesFromResponse(resp sttypes.Response) ([]common.Hash, error) {
+	sResp, ok := resp.(*syncResponse)
+	if !ok || sResp == nil {
+		return nil, errors.New("not sync response")
+	}
+	if errResp := sResp.pb.GetErrorResponse(); errResp != nil {
+		return nil, errors.New(errResp.Error)
+	}
+	bhResp := sResp.pb.GetGetBlockHashesResponse()
+	if bhResp == nil {
+		return nil, errors.New("response not GetBlockHashes")
+	}
+	hashBytes := bhResp.Hashes
+	return bytesToHashes(hashBytes), nil
+}
+
+type getBlocksByHashesRequest struct {
+	hashes []common.Hash
+	pbReq  *syncpb.Request
+}
+
+func newGetBlocksByHashesRequest(hashes []common.Hash) *getBlocksByHashesRequest {
+	pbReq := syncpb.MakeGetBlocksByHashesRequest(hashes)
+	return &getBlocksByHashesRequest{
+		hashes: hashes,
+		pbReq:  pbReq,
+	}
+}
+
+func (req *getBlocksByHashesRequest) ReqID() uint64 {
+	return req.pbReq.GetReqId()
+}
+
+func (req *getBlocksByHashesRequest) SetReqID(val uint64) {
+	req.pbReq.ReqId = val
+}
+
+func (req *getBlocksByHashesRequest) String() string {
+	hashStrs := make([]string, 0, len(req.hashes))
+	for _, h := range req.hashes {
+		hashStrs = append(hashStrs, fmt.Sprintf("%x", h[:]))
+	}
+	hStr := strings.Join(hashStrs, ", ")
+	return fmt.Sprintf("REQUEST [GetBlocksByHashes: %v]", hStr)
+}
+
+func (req *getBlocksByHashesRequest) IsSupportedByProto(target sttypes.ProtoSpec) bool {
+	return target.Version.GreaterThanOrEqual(MinVersion)
+}
+
+func (req *getBlocksByHashesRequest) Encode() ([]byte, error) {
+	msg := syncpb.MakeMessageFromRequest(req.pbReq)
+	return protobuf.Marshal(msg)
+}
+
+func (req *getBlocksByHashesRequest) getBlocksFromResponse(resp sttypes.Response) ([]*types.Block, error) {
+	sResp, ok := resp.(*syncResponse)
+	if !ok || sResp == nil {
+		return nil, errors.New("not sync response")
+	}
+	if errResp := sResp.pb.GetErrorResponse(); errResp != nil {
+		return nil, errors.New(errResp.Error)
+	}
+	bhResp := sResp.pb.GetGetBlocksByHashesResponse()
+	if bhResp == nil {
+		return nil, errors.New("response not GetBlocksByHashes")
+	}
+	var (
+		blockBytes = bhResp.BlocksBytes
+		sigs       = bhResp.CommitSig
+	)
+	if len(blockBytes) != len(sigs) {
+		return nil, fmt.Errorf("sig size not expected: %v / %v", len(sigs), len(blockBytes))
+	}
+	blocks := make([]*types.Block, 0, len(blockBytes))
+	for i, bb := range blockBytes {
+		var block *types.Block
+		if err := rlp.DecodeBytes(bb, &block); err != nil {
+			return nil, errors.Wrap(err, "[GetBlocksByHashesResponse]")
+		}
+		if block != nil {
+			block.SetCurrentCommitSig(sigs[i])
+		}
```

Comment:
what happens if block is nil?  seems you keep adding it?

Reply:
Yes. Nil block is added as a placeholder. The following equation is strictly defined in verification

```
len(req.bns) == len(resp.blockBytes) == len(resp.sigs)
```